Frameworks/AddressBook/ABAddressBook.mm

```diff
@@ -16,44 +16,87 @@
 
 #import <AddressBook/ABAddressBook.h>
 
-#import <StubReturn.h>
 #import "AssertARCEnabled.h"
+#import <StubReturn.h>
+
+#import "_ABAddressBook.h"
+
+#import "UWP/InteropBase.h"
+#import "UWP/WindowsDevicesEnumeration.h"
+#import "initguid.h"
+
+// GUID to check Windows Contacts permissions.
+DEFINE_GUID(_ABAddressBookContactsGUID, 0x7D7E8402, 0x7C54, 0x4821, 0xA3, 0x4E, 0xAE, 0xEF, 0xD6, 0x2D, 0xED, 0x93);
 
 const CFStringRef ABAddressBookErrorDomain = static_cast<const CFStringRef>(@"ABAddressBookErrorDomain");
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 ABAddressBookRef ABAddressBookCreate() {
-    UNIMPLEMENTED();
-    return StubReturn();
+    return ABAddressBookCreateWithOptions(NULL, NULL);
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 ABAddressBookRef ABAddressBookCreateWithOptions(CFDictionaryRef options, CFErrorRef* error) {
-    UNIMPLEMENTED();
-    return StubReturn();
+    _ABAddressBook* addressBook = [[_ABAddressBook alloc] init];
+    if (addressBook.contactStore == nil) {
+        // There was an error setting up the backing
+        // contact store, so just return NULL to signify
+        // the error.
+        if (error != NULL) {
+			// In this case, the caller is responsible to release error.
+            *error = CFErrorCreate(NULL, ABAddressBookErrorDomain, kABOperationNotPermittedByUserError, NULL);
+        }
+        return NULL;
+    } else {
+        return (__bridge_retained ABAddressBookRef)addressBook;
+    }
 }
 
 /**
- @Status Stub
+ @Status Interoperable
  @Notes
 */
 ABAuthorizationStatus ABAddressBookGetAuthorizationStatus() {
-    UNIMPLEMENTED();
-    return kABAuthorizationStatusNotDetermined;
+    WFGUID* guid = [WFGUID guidWithGUID:_ABAddressBookContactsGUID];
+    WDEDeviceAccessInformation* deviceAccessInformation = [WDEDeviceAccessInformation createFromDeviceClassId:guid];
+    WDEDeviceAccessStatus currentStatus = deviceAccessInformation.currentStatus;
+
+    if (currentStatus == WDEDeviceAccessStatusAllowed) {
+        return kABAuthorizationStatusAuthorized;
+    } else if (currentStatus == WDEDeviceAccessStatusDeniedBySystem) {
+        return kABAuthorizationStatusRestricted;
+    } else if (currentStatus == WDEDeviceAccessStatusDeniedByUser) {
+        return kABAuthorizationStatusDenied;
+    } else { // WDEDeviceAccessStatusUnspecified
+        return kABAuthorizationStatusNotDetermined;
+    }
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Caveat
+ @Notes Requesting access should be done by declaring Contacts capabilities in the app's package manifest -- it cannot be granted at
+ runtime.
 */
 void ABAddressBookRequestAccessWithCompletion(ABAddressBookRef addressBook, ABAddressBookRequestAccessCompletionHandler completion) {
```

Comment:
If the app doesn't have the capabilities added in their manifest, the privacy page won't show up in settings. We could deep-link in the case where an app had declared the contacts capability, but a user went out of their way to turn off the permissions in the settings app; however, the examples I've seen don't use this method to request permissions if a user has explicitly denied permission (they will only try to request permission if the status is not determined).


---
In reply to: [73075277](https://github.com/Microsoft/WinObjC/pull/786#discussion_r73075277) [](ancestors = 73075277)